web/src/queries/storage.ts

```diff
@@ -0,0 +1,349 @@
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, contact SUSE LLC.
+ *
+ * To contact SUSE LLC about this file by physical or electronic mail, you may
+ * find current contact information at www.suse.com.
+ */
+
+import {
+  useMutation,
+  useQuery,
+  useQueryClient,
+  useSuspenseQueries,
+  useSuspenseQuery,
+} from "@tanstack/react-query";
+import React from "react";
+import { fetchDevices, fetchDevicesDirty } from "~/api/storage/devices";
+import {
+  calculate,
+  fetchActions,
+  fetchDefaultVolume,
+  fetchProductParams,
+  fetchSettings,
+  fetchUsableDevices,
+} from "~/api/storage/proposal";
+import { useInstallerClient } from "~/context/installer";
+import { compact, uniq } from "~/utils";
+import {
+  ProductParams,
+  Volume as APIVolume,
+  ProposalSettings as APIProposalSettings,
+  ProposalTarget as APIProposalTarget,
+  ProposalSettingsPatch,
+} from "~/api/storage/types";
+import {
+  ProposalSettings,
+  ProposalResult,
+  ProposalTarget,
+  StorageDevice,
+  Volume,
+  VolumeTarget,
+} from "~/types/storage";
+
+const devicesQuery = (scope: "result" | "system") => ({
+  queryKey: ["storage", "devices", scope],
+  queryFn: () => fetchDevices(scope),
+  staleTime: Infinity,
+});
+
+const usableDevicesQuery = {
+  queryKey: ["storage", "usableDevices"],
+  queryFn: fetchUsableDevices,
+  staleTime: Infinity,
+};
+
+const productParamsQuery = {
+  queryKey: ["storage", "encryptionMethods"],
+  queryFn: fetchProductParams,
+  staleTime: Infinity,
+};
+
+const defaultVolumeQuery = (mountPath: string) => ({
+  queryKey: ["storage", "volumeFor", mountPath],
+  queryFn: () => fetchDefaultVolume(mountPath),
+  staleTime: Infinity,
+});
+
+/**
+ * Hook that returns the list of storage devices for the given scope.
+ *
+ * @param scope - "system": devices in the current state of the system; "result":
+ *   devices in the proposal ("stage")
+ */
+const useDevices = (
+  scope: "result" | "system",
+  options?: QueryHookOptions,
+): StorageDevice[] | undefined => {
+  const query = devicesQuery(scope);
+  const func = options?.suspense ? useSuspenseQuery : useQuery;
+  const { data } = func(query);
+  return data;
+};
+
+/**
+ * Hook that returns the list of available devices for installation.
+ */
+const useAvailableDevices = () => {
+  const findDevice = (devices: StorageDevice[], sid: number) => {
+    const device = devices.find((d) => d.sid === sid);
+
+    if (device === undefined) console.warn("Device not found:", sid);
+
+    return device;
+  };
+
+  const devices = useDevices("system", { suspense: true });
+  const { data } = useSuspenseQuery(usableDevicesQuery);
+
+  return data.map((sid) => findDevice(devices, sid)).filter((d) => d);
+};
+
+/**
+ * Hook that returns the product parameters (e.g., mount points).
+ */
+const useProductParams = (options?: QueryHookOptions): ProductParams => {
```

Comment:
Interesting... should it be here or in the software queries?

Reply:
IMHO it belongs here. They are storage-related params that depend on the product.